services/metric_service.go

```diff
@@ -266,3 +277,45 @@ func (m MetricService) Send(ctx context.Context, offline bool, metricPostDuratio
 		}
 	}()
 }
+
+// subscribe kicks off a job that subscribes to the redis stream that readreplicas write metrics onto
+func (m MetricService) subscribe(ctx context.Context) {
+	id := ""
+	for {
+		// There's nothing we can really do here if we error in the callback parsing/handling the message
+		// so we log the errors as warnings but return nil so we carry on receiving messages from the stream
+		err := m.subscriber.Sub(ctx, SDKMetricsStream, id, func(latestID string, v interface{}) error {
+			// We want to keep track of the id of the latest message we've received so that if
+			// we disconnect we can resume from that point
+			id = latestID
+
+			s, ok := v.(string)
+			if !ok {
+				m.log.Warn("unexpected message format received", "stream", SDKMetricsStream, "type", reflect.TypeOf(v))
+				return nil
+			}
+
+			mr := domain.MetricsRequest{}
+			if err := jsoniter.Unmarshal([]byte(s), &mr); err != nil {
+				m.log.Warn("failed to unmarshal metrics message", "stream", SDKMetricsStream, "err", err)
+				return nil
+			}
+
+			if err := m.StoreMetrics(ctx, mr); err != nil {
+				m.log.Warn("failed to store metrics received from read replica", "stream", SDKMetricsStream, "err", err)
+				return nil
+			}
+			return nil
+		})
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				return
+			}
+
+			// If we do break out of subscribe it will probably be because of a connection error with redis
+			// so backoff for 30 seconds before trying again
+			time.Sleep(30 * time.Second)
```

Comment:
Would it be worth adding a log here, to give an indication we lost connection and will reconnect?  If it recovers its not a big deal, but if we can never reconnect it would point to an issue.

Reply:
Yea good shout